protocols/v2/noise-sv2/src/responder.rs

```diff
@@ -153,36 +160,72 @@ impl Responder {
     ///
-    /// Message length: 170 bytes
-    pub fn step_1(&mut self, re_pub: [u8; 32]) -> Result<([u8; 170], NoiseCodec), aes_gcm::Error> {
+    /// Message length: 232 bytes
+    pub fn step_1(
+        &mut self,
+        elligatorswift_theirs_ephemeral_serialized: [u8; ELLSWIFT_ENCODING_SIZE],
+    ) -> Result<
+        (
+            [u8; INITIATOR_EXPECTED_HANDSHAKE_MESSAGE_LENGTH],
+            NoiseCodec,
+        ),
+        aes_gcm::Error,
+    > {
         // 4.5.1.2 Responder
-        Self::mix_hash(self, &re_pub[..]);
+        Self::mix_hash(self, &elligatorswift_theirs_ephemeral_serialized[..]);
         Self::decrypt_and_hash(self, &mut vec![])?;
 
         // 4.5.2.1 Responder
-        let mut out = [0; 170];
-        let serialized = self.e.x_only_public_key().0.serialize();
-        out[..32].copy_from_slice(&serialized[..32]);
+        let mut out = [0; INITIATOR_EXPECTED_HANDSHAKE_MESSAGE_LENGTH];
+        let keypair = self.e;
+        let elligatorswitf_ours_ephemeral = ElligatorSwift::from_pubkey(keypair.public_key());
+        let elligatorswift_ours_ephemeral_serialized = elligatorswitf_ours_ephemeral.to_array();
+        out[..ELLSWIFT_ENCODING_SIZE]
+            .copy_from_slice(&elligatorswift_ours_ephemeral_serialized[..ELLSWIFT_ENCODING_SIZE]);
 
         // 3. calls `MixHash(e.public_key)`
-        Self::mix_hash(self, &serialized);
+        // what is here is not the public key encoded with ElligatorSwift, but the x-coordinate of
+        // the public key (which is a point in the EC).
+
+        Self::mix_hash(self, &elligatorswift_ours_ephemeral_serialized);
 
         // 4. calls `MixKey(ECDH(e.private_key, re.public_key))`
-        let e_private_key = self.e.secret_bytes();
-        let ecdh = Self::ecdh(&e_private_key[..], &re_pub[..]);
-        Self::mix_key(self, &ecdh);
+        let e_private_key = keypair.secret_key();
+        let elligatorswift_theirs_ephemeral =
+            ElligatorSwift::from_array(elligatorswift_theirs_ephemeral_serialized);
+        let ecdh_ephemeral = ElligatorSwift::shared_secret(
+            elligatorswift_theirs_ephemeral,
+            elligatorswitf_ours_ephemeral,
+            e_private_key,
+            secp256k1::ellswift::ElligatorSwiftParty::B,
+            None,
+        )
+        .to_secret_bytes();
+        Self::mix_key(self, &ecdh_ephemeral);
 
-        // 5. appends `EncryptAndHash(s.public_key)` (32 bytes encrypted public key, 16 bytes MAC)
-        let mut encrypted_static_pub_k = vec![0; 32];
-        let static_pub_k = self.s.x_only_public_key().0.serialize();
-        encrypted_static_pub_k[..32].copy_from_slice(&static_pub_k[..32]);
+        // 5. appends `EncryptAndHash(s.public_key)` (64 bytes encrypted elligatorswift  public key, 16 bytes MAC)
+        let mut encrypted_static_pub_k = vec![0; ELLSWIFT_ENCODING_SIZE];
+        let elligatorswift_ours_static = ElligatorSwift::from_pubkey(self.s.public_key());
```

Comment:
It is possible to get the EllSwift encoded key directly from the private key?

That might be a bit faster. Though performance is not a big deal here.

Reply:
Yes we can derive the Ellswift pubkey from the private key. Moreover, in the SRI internal representation of Initiator and Responder handshake roles it is possible to store only the private key and derive whatever we want from that. This goes even further. 
I am not totally against it, but I would prefer to convert the public key instead  for two reasons
1. because it resembles more the idea that then Ellswift is just an encoding
2. I would like to think about it, test it and generally be more conservative an proceed by steps

What do you think about?

Reply:
I don't have a strong preference.